src/core/optimize.ts

```diff
@@ -1,43 +1,86 @@
+import { assert } from "./assert"
 import * as e from "./eval"
 import * as g from "./grad"
-import { wasmOptimizer } from "./wasmopt"
+import { collectParams } from "./params"
 import * as t from "./types"
+import { wasmOptimizer } from "./wasmopt"
 
 const useWasm = true
 
-export function optimize(
+export interface Optimizer {
+  optimize(iterations: number, observations?: Map<t.Param, number>): e.Evaluator
+}
+
+function basicOptimizer(
   loss: t.Num,
+  gradient: Map<t.Param, t.Num>,
   init: Map<t.Param, number>,
-  iterations: number,
-): e.Evaluator {
+) {
+  return (iterations: number, observations: Map<t.Param, number>) => {
+    const params = new Map([...init, ...observations])
+    const epsilon = 0.0001
+    let i = iterations
+    while (i > 0) {
+      const ev = e.evaluator(params)
+      const l = ev.evaluate(loss)
+      if (i % 1000 == 0) {
+        console.log(l)
+      }
+      gradient.forEach((v, k) => {
+        const diff = ev.evaluate(v)
+        const old = params.get(k) || 0
+        const update = old - diff * epsilon
+        params.set(k, update)
+      })
+      i = i - 1
+    }
+    return new Map(params.entries())
+  }
+}
+
+export function optimizer(loss: t.Num, init?: Map<t.Param, number>): Optimizer {
   const gradient = g.gradient(loss)
-  const params = new Map(init)
+
+  // Ensure that we have an initial value for all free parameters.
+  const freeParams = new Map(init)
   gradient.forEach((_, k) => {
-    if (!params.has(k)) params.set(k, Math.random() * 10)
+    if (!freeParams.has(k)) freeParams.set(k, Math.random() * 10)
```

Comment:
while I'm looking at this: we should use a standard normal to initialize the params, eg, 

```
  Math.sqrt( -2 * Math.log(1 - Math.random()) ) * Math.cos( 2 * Math.PI * Math.random() )
```